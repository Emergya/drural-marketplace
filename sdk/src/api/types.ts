import { ApolloError } from "apollo-client";

import { PasswordChange } from "../mutations/gqlTypes/PasswordChange";
import { SetPassword } from "../mutations/gqlTypes/SetPassword";
import { TokenAuth_tokenCreate } from "../mutations/gqlTypes/TokenAuth";
import BaseDetails from "../helpers/BaseDetails";
import BaseList, { BaseListVariables } from "../helpers/BaseList";

export interface ErrorResponse<T> {
  error?: any;
  type?: T;
}

export interface FunctionQueueResponse {
  pending: boolean;
}
export interface FunctionRunResponse<D, F> {
  data?: any;
  dataError?: ErrorResponse<D>;
  functionError?: ErrorResponse<F>;
  pending?: boolean;
}

export type PromiseQueuedResponse = Promise<FunctionQueueResponse>;
export type PromiseRunResponse<D, F = undefined> = Promise<
  FunctionRunResponse<D, F>
>;

export type SignIn = {
  data: TokenAuth_tokenCreate | null;
  error: ApolloError | null;
} | null;

export type SetPasswordChange = {
  data: PasswordChange | null;
  error: ApolloError | null;
} | null;

export type SetPasswordResult = {
  data: SetPassword | null;
  error: ApolloError | null;
} | null;

export interface WithDetails<TQuery, TObject, TVariables> {
  getDetails(
    variables: TVariables
  ): Promise<BaseDetails<TQuery, TObject, TVariables>>;
}

export interface WithList<
  TQuery,
  TObject,
  TVariables extends BaseListVariables
> {
  getList(
    variables: TVariables
  ): Promise<BaseList<TQuery, TObject, TVariables>>;
}
