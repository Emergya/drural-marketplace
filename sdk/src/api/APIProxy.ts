import {
  ApolloClient,
  ApolloError,
  ObservableQuery,
  WatchQueryOptions,
} from "apollo-client";
import { GraphQLError } from "graphql";

import { UserOrderByToken } from "../queries/gqlTypes/UserOrderByToken";
import { OrderByToken } from "../queries/gqlTypes/OrderByToken";
import { PasswordChange } from "../mutations/gqlTypes/PasswordChange";
import { SetPassword } from "../mutations/gqlTypes/SetPassword";
import { getAuthToken } from "../auth";
import { MUTATIONS } from "../mutations";
import { QUERIES } from "../queries";
import { RequireAtLeastOne } from "../tsHelpers";
import {
  InferOptions,
  MapFn,
  QueryShape,
  WatchMapFn,
  WatchQueryData,
} from "../types";
import {
  getErrorsFromData,
  getMappedData,
  isDataEmpty,
  mergeEdges,
} from "../utils";
import { SetPasswordChange, SetPasswordResult } from "./types";
import { WINDOW_EXISTS } from "../consts";

const handleDataErrors = <T extends QueryShape, TData>(
  mapFn: MapFn<T, TData> | WatchMapFn<T, TData>,
  data: TData,
  apolloErrors?: readonly GraphQLError[]
) => {
  // INFO: user input errors will be moved to graphql errors
  const userInputErrors = getErrorsFromData(data);
  const errors =
    apolloErrors || userInputErrors
      ? new ApolloError({
          extraInfo: userInputErrors,
          graphQLErrors: apolloErrors,
        })
      : null;

  if (errors && isDataEmpty(data)) {
    return { errors };
  }

  const result = getMappedData(mapFn, data);

  return { data: result };
};

class APIProxy {
  client: ApolloClient<any>;

  constructor(client: ApolloClient<any>) {
    this.client = client;
  }

  getAttributes = this.watchQuery(QUERIES.Attributes, data => data.attributes);

  getOrdersByUser = this.watchQuery(QUERIES.OrdersByUser, data =>
    data.me ? data.me.orders : null
  );

  getOrderDetails = (
    variables: InferOptions<
      QUERIES["OrderDetails"] | QUERIES["OrderDetailsByUser"]
    >["variables"],
    options: Omit<
      InferOptions<QUERIES["OrderDetails"] | QUERIES["OrderDetailsByUser"]>,
      "variables"
    > & {
      onUpdate: (
        data:
          | OrderByToken["orderByToken"]
          | UserOrderByToken["orderByToken"]
          | null
      ) => void;
    }
  ) => {
    if (this.isLoggedIn()) {
      return this.watchQuery(
        QUERIES.OrderDetailsByUser,
        data => data.orderByToken
      )(variables, options);
    }
    return this.watchQuery(QUERIES.OrderDetails, data => data.orderByToken)(
      variables,
      options
    );
  };

  getVariantsProducts = this.watchQuery(
    QUERIES.VariantsProducts,
    data => data.productVariants
  );

  getShopDetails = this.watchQuery(QUERIES.GetShopDetails, data => data);

  setUserDefaultAddress = this.fireQuery(
    MUTATIONS.AddressTypeUpdate,
    data => data!.accountSetDefaultAddress
  );

  setDeleteUserAddress = this.fireQuery(
    MUTATIONS.DeleteUserAddress,
    data => data!.accountAddressDelete
  );

  setCreateUserAddress = this.fireQuery(
    MUTATIONS.CreateUserAddress,
    data => data!.accountAddressCreate
  );

  setUpdateuserAddress = this.fireQuery(
    MUTATIONS.UpdateUserAddress,
    data => data!.accountAddressUpdate
  );

  setAccountUpdate = this.fireQuery(
    MUTATIONS.AccountUpdate,
    data => data!.accountUpdate
  );

  setPassword = async (
    variables: InferOptions<MUTATIONS["SetPassword"]>["variables"],
    options?: Omit<InferOptions<MUTATIONS["SetPassword"]>, "variables">
  ): Promise<SetPasswordResult> => {
    let result: {
      data: SetPassword | null;
    } | null = null;

    result = await this.fireQuery(MUTATIONS.SetPassword, data => data!)(
      variables,
      {
        ...options,
        fetchPolicy: "no-cache",
      }
    );
    const { data } = result;

    return {
      data,
      error: null,
    };
  };

  setPasswordChange = async (
    variables: InferOptions<MUTATIONS["PasswordChange"]>["variables"],
    options?: Omit<InferOptions<MUTATIONS["PasswordChange"]>, "variables">
  ): Promise<SetPasswordChange> => {
    let result: {
      data: PasswordChange | null;
    } | null = null;

    result = await this.fireQuery(MUTATIONS.PasswordChange, data => data!)(
      variables,
      {
        ...options,
        fetchPolicy: "no-cache",
      }
    );
    const { data } = result;

    return {
      data,
      error: null,
    };
  };

  attachAuthListener = (callback: (authenticated: boolean) => void) => {
    const eventHandler = () => {
      callback(this.isLoggedIn());
    };

    if (WINDOW_EXISTS) {
      window.addEventListener("auth", eventHandler);
    }

    return () => {
      if (WINDOW_EXISTS) {
        window.removeEventListener("auth", eventHandler);
      }
    };
  };

  isLoggedIn = () => {
    return !!getAuthToken();
  };

  watchQuery<T extends QueryShape, TResult>(
    query: T,
    mapFn: WatchMapFn<T, TResult>
  ) {
    return <
      TVariables extends InferOptions<T>["variables"],
      TOptions extends Omit<
        InferOptions<T> | WatchQueryOptions<InferOptions<T>>,
        "variables"
      >
    >(
      variables: TVariables,
      options: TOptions & {
        skip?: boolean;
        onComplete?: () => void;
        onError?: (error: ApolloError) => void;
        onUpdate: (
          data: ReturnType<typeof mapFn> | null,
          loading?: boolean
        ) => void;
      }
    ) => {
      const { onComplete, onError, onUpdate, ...apolloClientOptions } = options;

      const observable: ObservableQuery<WatchQueryData<T>, TVariables> = query(
        this.client,
        {
          ...apolloClientOptions,
          variables,
        }
      );

      if (options.skip) {
        return {
          refetch: () => {
            return new Promise(resolve => {
              resolve({ data: null });
            });
          },
          unsubscribe: null,
        };
      }

      const subscription = observable.subscribe(
        result => {
          const { data, errors: apolloErrors, loading } = result;
          const errorHandledData = handleDataErrors(
            mapFn,
            data as any,
            apolloErrors
          );
          if (onUpdate) {
            if (errorHandledData.errors) {
              if (onError) {
                onError(errorHandledData.errors);
              }
            } else {
              onUpdate(errorHandledData.data as TResult, loading);
              if (onComplete) {
                onComplete();
              }
            }
          }
        },
        error => {
          if (onError) {
            onError(error);
          }
        }
      );

      return {
        loadMore: (
          extraVariables: RequireAtLeastOne<TVariables>,
          mergeResults: boolean = true
        ) => {
          observable.fetchMore({
            updateQuery: (previousResult, { fetchMoreResult }) => {
              if (!fetchMoreResult) {
                // returning previousResult doesn't trigger observable `next`
                onUpdate(mapFn(previousResult));
                return previousResult;
              }

              if (mergeResults) {
                const prevResultRef = mapFn(previousResult) as any;
                const newResultRef = mapFn(fetchMoreResult) as any;

                if (!prevResultRef || !newResultRef) {
                  onUpdate(prevResultRef);
                  return previousResult;
                }

                const mergedEdges = mergeEdges(
                  prevResultRef.edges,
                  newResultRef.edges
                );

                // use new result for metadata and mutate existing data
                Object.keys(prevResultRef).forEach(key => {
                  prevResultRef[key] = newResultRef[key];
                });
                prevResultRef.edges = mergedEdges;

                return previousResult;
              }

              return fetchMoreResult;
            },
            variables: { ...variables, ...extraVariables },
          });
        },
        refetch: (newVariables?: TVariables) => {
          if (newVariables) {
            observable.setVariables(newVariables);
            const cachedResult = observable.currentResult();
            const errorHandledData = handleDataErrors(mapFn, cachedResult.data);
            if (errorHandledData.data) {
              onUpdate(errorHandledData.data as TResult);
            }
          }

          return APIProxy.firePromise(
            () => observable.refetch(newVariables),
            mapFn
          );
        },
        setOptions: (newOptions: TOptions) =>
          APIProxy.firePromise(() => observable.setOptions(newOptions), mapFn),
        unsubscribe: subscription.unsubscribe.bind(subscription),
      };
    };
  }

  fireQuery<T extends QueryShape, TResult>(query: T, mapFn: MapFn<T, TResult>) {
    return (
      variables: InferOptions<T>["variables"],
      options?: Omit<InferOptions<T>, "variables">
    ) =>
      APIProxy.firePromise(
        () =>
          query(this.client, {
            ...options,
            variables,
          }),
        mapFn
      );
  }

  // Promise wrapper to catch errors
  static firePromise<T extends QueryShape, TResult>(
    promise: () => Promise<any>,
    mapFn: MapFn<T, TResult> | WatchMapFn<T, TResult>
  ) {
    return new Promise<{ data: ReturnType<typeof mapFn> | null }>(
      async (resolve, reject) => {
        try {
          const { data, errors: apolloErrors } = await promise();
          const errorHandledData = handleDataErrors(mapFn, data, apolloErrors);

          if (errorHandledData.errors) {
            reject(errorHandledData.errors);
          }

          resolve({ data: errorHandledData.data });
        } catch (error) {
          reject(error);
        }
      }
    );
  }
}

export default APIProxy;
