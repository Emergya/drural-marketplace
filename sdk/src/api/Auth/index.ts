import { User } from "../../fragments/gqlTypes/User";
import { ErrorListener } from "../../helpers";
import { JobsManager } from "../../jobs";
import { SaleorState, SaleorStateLoaded } from "../../state";
import { StateItems } from "../../state/types";

import { PromiseRunResponse } from "../types";
import { DataErrorAuthTypes } from "./types";
import { Config } from "../../types";
import { CREDENTIAL_API_EXISTS } from "../../consts";

export const BROWSER_NO_CREDENTIAL_API_MESSAGE =
  "Saleor SDK is unable to use browser Credential Management API.";

export class AuthAPI extends ErrorListener {
  /**
   * Indicates if data is initialized, initially retrieved from cache or initially fetched.
   */
  loaded: boolean;

  /**
   * User object with currently signed in user data.
   */
  user?: User | null;

  /**
   * Indicates if user is signed in.
   */
  authenticated?: boolean;

  /**
   * Token used for user authentication.
   */
  token?: string;

  /**
   * Indicate if token refreshing is in progress.
   */
  tokenRefreshing: boolean;

  /**
   * Indicate if token verifying is in progress.
   */
  tokenVerifying: boolean;

  private saleorState: SaleorState;

  private jobsManager: JobsManager;

  // temporary solution, might change in future
  private config: Config;

  constructor(
    saleorState: SaleorState,
    jobsManager: JobsManager,
    config: Config
  ) {
    super();
    this.saleorState = saleorState;
    this.jobsManager = jobsManager;
    this.config = config;

    this.loaded = false;
    this.tokenRefreshing = false;
    this.tokenVerifying = !!this.saleorState.signInToken;

    this.saleorState.subscribeToChange(StateItems.USER, (user: User | null) => {
      this.user = user;
      if (this.loaded) {
        this.authenticated = !!this.user;
      }
    });
    this.saleorState.subscribeToChange(StateItems.SIGN_IN_TOKEN, token => {
      this.token = token;
    });
    this.saleorState.subscribeToChange(
      StateItems.SIGN_IN_TOKEN_REFRESHING,
      tokenRefreshing => {
        this.tokenRefreshing = tokenRefreshing;
      }
    );
    this.saleorState.subscribeToChange(
      StateItems.SIGN_IN_TOKEN_VERIFYING,
      tokenVerifying => {
        this.tokenVerifying = tokenVerifying;
      }
    );
    this.saleorState.subscribeToChange(
      StateItems.LOADED,
      (loaded: SaleorStateLoaded) => {
        this.loaded = loaded.user && loaded.signInToken;
        if (this.loaded) {
          this.authenticated = !!this.user;
        }
      }
    );

    if (!this.saleorState.signInToken && CREDENTIAL_API_EXISTS) {
      this.autoSignIn();
    }
  }

  /**
   * Tries to register a user account with given email and password.
   * @param email Email used for new account.
   * @param password Password used for new account.
   * @param redirectUrl URL used for redirection.
   */
  registerAccount = async (
    email: string,
    password: string,
    redirectUrl: string
  ): PromiseRunResponse<DataErrorAuthTypes> => {
    const { data, dataError } = await this.jobsManager.run(
      "auth",
      "registerAccount",
      {
        email,
        password,
        redirectUrl,
      }
    );

    if (dataError?.error) {
      this.fireError(dataError.error, DataErrorAuthTypes.REGISTER_ACCOUNT);
    }

    if (dataError) {
      return {
        data,
        dataError,
        pending: false,
      };
    }

    return {
      data,
      pending: false,
    };
  };

  /**
   * Requests a password reset for an user account with given email.
   * @param email Email used for account.
   * @param redirectUrl URL used for redirection.
   */
  resetPasswordRequest = async (
    email: string,
    redirectUrl: string
  ): PromiseRunResponse<DataErrorAuthTypes> => {
    const { data, dataError } = await this.jobsManager.run(
      "auth",
      "resetPasswordRequest",
      {
        email,
        redirectUrl,
      }
    );

    if (dataError?.error) {
      this.fireError(
        dataError.error,
        DataErrorAuthTypes.RESET_PASSWORD_REQUEST
      );
    }

    if (dataError) {
      return {
        data,
        dataError,
        pending: false,
      };
    }

    return {
      data,
      pending: false,
    };
  };

  /**
   * Tries to authenticate user with given email and password.
   * @param email Email used for authentication.
   * @param password Password used for authentication.
   * @param autoSignIn Indicates if SDK should try to sign in user with given credentials in future without explicitly calling this method. True by default.
   */
  signIn = async (
    email: string,
    password: string,
    autoSignIn: boolean = true
  ): PromiseRunResponse<DataErrorAuthTypes> => {
    const { data, dataError } = await this.jobsManager.run("auth", "signIn", {
      email,
      password,
    });

    try {
      if (autoSignIn && !dataError?.error && CREDENTIAL_API_EXISTS) {
        await navigator.credentials.store(
          new window.PasswordCredential({
            id: email,
            password,
          })
        );
      }
    } catch (credentialsError) {
      // eslint-disable-next-line no-console
      console.warn(BROWSER_NO_CREDENTIAL_API_MESSAGE, credentialsError);
    }

    if (dataError) {
      return {
        data,
        dataError,
        pending: false,
      };
    }

    const {
      data: userData,
      dataError: userDataError,
    } = await this.jobsManager.run("auth", "provideUser", undefined);
    if (this.config.loadOnStart.checkout) {
      await this.jobsManager.run("checkout", "provideCheckout", {
        channel: this.config.channel,
        isUserSignedIn: !!data?.user,
      });
    }

    return {
      data: userData,
      dataError: userDataError,
      pending: false,
    };
  };

  /**
   * Updates user data
   */

  refreshUserData = async () => {
    const {
      data: userData,
      dataError: userDataError,
    } = await this.jobsManager.run("auth", "provideUser", undefined);

    return {
      data: userData,
      dataError: userDataError,
      pending: false,
    };
  };

  /**
   * Sign out user by clearing cache, local storage and authentication token.
   */
  signOut = async (): PromiseRunResponse<DataErrorAuthTypes> => {
    await this.jobsManager.run("auth", "signOut", undefined);
    try {
      if (navigator.credentials?.preventSilentAccess) {
        await navigator.credentials.preventSilentAccess();
      }
    } catch (credentialsError) {
      // eslint-disable-next-line no-console
      console.warn(BROWSER_NO_CREDENTIAL_API_MESSAGE, credentialsError);
    }

    return {
      pending: false,
    };
  };

  /**
   * Tries to refresh user token to keep previously signed in user authenticated.
   * @param refreshToken Refresh token. Required when refreshToken is not provided as a cookie.
   */
  refreshSignInToken = async (
    refreshToken?: string
  ): PromiseRunResponse<DataErrorAuthTypes> => {
    const { data, dataError } = await this.jobsManager.run(
      "auth",
      "refreshSignInToken",
      {
        refreshToken,
      }
    );

    if (dataError) {
      return {
        data,
        dataError,
      };
    }

    return {
      data,
    };
  };

  private autoSignIn = async () => {
    let credentials;
    try {
      credentials = await navigator.credentials.get({
        password: true,
      });
    } catch (credentialsError) {
      // eslint-disable-next-line no-console
      console.warn(BROWSER_NO_CREDENTIAL_API_MESSAGE, credentialsError);
    }

    if (credentials && "password" in credentials && credentials.password) {
      const { dataError } = await this.signIn(
        credentials.id,
        credentials.password,
        true
      );

      if (dataError?.error) {
        this.fireError(dataError.error, DataErrorAuthTypes.SIGN_IN);
      }
    }
  };
}
