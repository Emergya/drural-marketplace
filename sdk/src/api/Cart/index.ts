import { LocalStorageManager } from "../../data";
import { ErrorListener } from "../../helpers";
import { ICheckoutModel } from "../../helpers/LocalStorageHandler";
import { JobsManager } from "../../jobs";
import { ErrorCartTypes } from "../../jobs/Cart";
import { SaleorState, SaleorStateLoaded } from "../../state";
import { ISaleorStateSummeryPrices, StateItems } from "../../state/types";
import { ApolloClientManager } from "../../data/ApolloClientManager";
import { sortCheckoutLines } from "./utils";

import { Config } from "../../types";
import {
  IDiscount,
  IItems,
  IShippingPrice,
  ISubtotalPrice,
  ITotalPrice,
} from "./types";

export class SaleorCartAPI extends ErrorListener {
  loaded: boolean;

  items: IItems;

  totalPrice: ITotalPrice;

  subtotalPrice: ISubtotalPrice;

  shippingPrice: IShippingPrice;

  discount?: IDiscount;

  private apolloClientManager: ApolloClientManager;

  private jobsManager: JobsManager;

  private localStorageManager: LocalStorageManager;

  private saleorState: SaleorState;

  // temporary solution, might change in future
  private config: Config;

  constructor(
    localStorageManager: LocalStorageManager,
    apolloClientManager: ApolloClientManager,
    saleorState: SaleorState,
    jobsManager: JobsManager,
    config: Config
  ) {
    super();
    this.saleorState = saleorState;
    this.localStorageManager = localStorageManager;
    this.apolloClientManager = apolloClientManager;
    this.jobsManager = jobsManager;
    this.config = config;

    this.loaded = false;

    this.jobsManager.attachErrorListener("cart", this.fireError);

    this.saleorState.subscribeToChange(
      StateItems.CHECKOUT,
      (checkout: ICheckoutModel) => {
        this.items = checkout?.lines
          ?.filter(line => line.quantity > 0)
          .sort(sortCheckoutLines);
      }
    );
    this.saleorState.subscribeToChange(
      StateItems.SUMMARY_PRICES,
      (summaryPrices: ISaleorStateSummeryPrices) => {
        const { totalPrice, subtotalPrice, shippingPrice, discount } =
          summaryPrices || {};
        this.totalPrice = totalPrice;
        this.subtotalPrice = subtotalPrice;
        this.shippingPrice = shippingPrice;
        this.discount = discount;
      }
    );
    this.saleorState.subscribeToChange(
      StateItems.LOADED,
      (loaded: SaleorStateLoaded) => {
        this.loaded = loaded.checkout && loaded.summaryPrices;
      }
    );
  }

  addItem = async (variantId: string, quantity: number) => {
    // 1. save in local storage
    this.localStorageManager.addItemToCart(variantId, quantity);

    // 2. save online if possible (if checkout id available)
    if (this.saleorState.checkout?.lines) {
      const {
        data,
        error,
      } = await this.apolloClientManager.getRefreshedCheckoutLines(
        this.saleorState.checkout.lines,
        this.config.channel
      );

      if (error) {
        this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
      } else {
        this.localStorageManager.getHandler().setCheckout({
          ...this.saleorState.checkout,
          lines: data,
        });
      }
    }

    if (this.saleorState.checkout?.id) {
      this.jobsManager.addToQueue("cart", "setCartItem");
      return {
        pending: true,
      };
    }
    return {
      pending: false,
    };
  };

  removeItem = async (variantId: string) => {
    // 1. save in local storage
    this.localStorageManager.removeItemFromCart(variantId);
    // 2. save online if possible (if checkout id available)
    if (this.saleorState.checkout?.lines) {
      const {
        data,
        error,
      } = await this.apolloClientManager.getRefreshedCheckoutLines(
        this.saleorState.checkout.lines,
        this.config.channel
      );

      if (error) {
        this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
      } else {
        this.localStorageManager.getHandler().setCheckout({
          ...this.saleorState.checkout,
          lines: data,
        });
      }
    }
    if (this.saleorState.checkout?.id) {
      this.jobsManager.addToQueue("cart", "setCartItem");
      return {
        pending: true,
      };
    }
    return {
      pending: false,
    };
  };

  subtractItem = async (variantId: string) => {
    // 1. save in local storage
    this.localStorageManager.subtractItemFromCart(variantId);

    // 2. save online if possible (if checkout id available)
    if (this.saleorState.checkout?.lines) {
      const {
        data,
        error,
      } = await this.apolloClientManager.getRefreshedCheckoutLines(
        this.saleorState.checkout.lines,
        this.config.channel
      );

      if (error) {
        this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
      } else {
        this.localStorageManager.getHandler().setCheckout({
          ...this.saleorState.checkout,
          lines: data,
        });
      }
    }
    if (this.saleorState.checkout?.id) {
      this.jobsManager.addToQueue("cart", "setCartItem");
      return {
        pending: true,
      };
    }
    return {
      pending: false,
    };
  };

  updateItem = async (
    variantId: string,
    quantity: number,
    bookingId?: string
  ) => {
    // 1. save in local storage
    this.localStorageManager.updateItemInCart(variantId, quantity, bookingId);

    // 2. save online if possible (if checkout id available)
    if (this.saleorState.checkout?.lines) {
      const {
        data,
        error,
      } = await this.apolloClientManager.getRefreshedCheckoutLines(
        this.saleorState.checkout.lines,
        this.config.channel
      );

      if (error) {
        this.fireError(error, ErrorCartTypes.SET_CART_ITEM);
      } else {
        this.localStorageManager.getHandler().setCheckout({
          ...this.saleorState.checkout,
          lines: data,
        });
      }
    }

    if (this.saleorState.checkout?.id) {
      this.jobsManager.addToQueue("cart", "setCartItem");
      return {
        pending: true,
      };
    }
    return {
      pending: false,
    };
  };
}
