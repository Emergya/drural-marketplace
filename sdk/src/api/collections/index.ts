export * from "./collections";
export * from "./CollectionList";
