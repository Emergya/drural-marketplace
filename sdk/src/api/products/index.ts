export * from "./products";
export * from "./ProductList";
