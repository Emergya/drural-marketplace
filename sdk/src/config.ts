import { DefaultConfig } from "./types";

export const defaultConfig: DefaultConfig = {
  channel: "",
  loadOnStart: {
    auth: true,
    checkout: true,
  },
};
