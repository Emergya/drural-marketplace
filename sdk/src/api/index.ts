import ApolloClient from "apollo-client";

import { LocalStorageManager } from "../data";
import { ApolloClientManager } from "../data/ApolloClientManager";
import { LocalStorageHandler } from "../helpers/LocalStorageHandler";
import { JobsManager } from "../jobs";
import { SaleorState } from "../state";
import { Config } from "../types";
import { AuthAPI } from "./Auth";
import APIProxy from "./APIProxy";
import { SaleorCartAPI } from "./Cart";
import { SaleorCheckoutAPI } from "./Checkout";
import { CollectionsAPI } from "./collections/collections";
import { CategoriesAPI } from "./categories/categories";
import { ProductsAPI } from "./products/products";

export * from "./Checkout";
export * from "./Cart";

export class SaleorAPI {
  auth: AuthAPI;

  checkout: SaleorCheckoutAPI;

  cart: SaleorCartAPI;

  categories: CategoriesAPI;

  collections: CollectionsAPI;

  products: ProductsAPI;

  /**
   * @deprecated Please do not use it anymore. Reference to API Proxy will be removed in future.
   * Now it just exists for legacy React hooks, which also will be removed.
   */
  legacyAPIProxy: APIProxy;

  constructor(
    client: ApolloClient<any>,
    apiProxy: APIProxy,
    config: Config,
    onStateUpdate?: () => any
  ) {
    this.legacyAPIProxy = apiProxy;

    const localStorageHandler = new LocalStorageHandler();
    const apolloClientManager = new ApolloClientManager(client);
    const jobsManager = new JobsManager(
      localStorageHandler,
      apolloClientManager
    );
    const saleorState = new SaleorState(
      config,
      localStorageHandler,
      apolloClientManager,
      jobsManager
    );
    const localStorageManager = new LocalStorageManager(
      localStorageHandler,
      saleorState
    );

    if (onStateUpdate) {
      saleorState.subscribeToNotifiedChanges(onStateUpdate);
    }

    this.auth = new AuthAPI(saleorState, jobsManager, config);
    this.checkout = new SaleorCheckoutAPI(saleorState, jobsManager, config);
    this.cart = new SaleorCartAPI(
      localStorageManager,
      apolloClientManager,
      saleorState,
      jobsManager,
      config
    );
    this.categories = new CategoriesAPI(client);
    this.collections = new CollectionsAPI(client, config);
    this.products = new ProductsAPI(client, config);

    this.legacyAPIProxy.attachAuthListener(authenticated => {
      if (!authenticated) {
        localStorageHandler.setCheckout({});
        localStorageHandler.setPayment({});
        localStorageHandler.setJobs(null);
      }
    });
  }
}
