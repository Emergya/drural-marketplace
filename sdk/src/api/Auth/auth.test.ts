import { InMemoryCache } from "apollo-cache-inmemory";
import { setupRecording, setupAPI } from "../../../testUtils/api";
import { AuthAPI } from ".";
import { defaultConfig } from "../../config";
import { LocalStorageHandler } from "../../helpers";
import { ApolloClientManager } from "../../data/ApolloClientManager";
import { JobsManager } from "../../jobs";
import { SaleorState } from "../../state";
import APIProxy from "../APIProxy";

setupRecording();

describe("Auth API", () => {
  let authAPI: AuthAPI;
  let cache: InMemoryCache;
  let apiProxy: APIProxy;

  beforeAll(async done => {
    const { client, cache: apiCache, apiUrl } = await setupAPI();
    cache = apiCache;

    const config = {
      ...defaultConfig,
      apiUrl,
      channel: "default-channel",
    };
    const localStorageHandler = new LocalStorageHandler();
    const apolloClientManager = new ApolloClientManager(client);
    const jobsManager = new JobsManager(
      localStorageHandler,
      apolloClientManager
    );
    const saleorState = new SaleorState(
      config,
      localStorageHandler,
      apolloClientManager,
      jobsManager
    );

    authAPI = new AuthAPI(saleorState, jobsManager, config);

    apiProxy = new APIProxy(client);

    done();
  });

  it("Returns error if credentials are invalid", async () => {
    const signInResult = await authAPI.signIn("sdk@example.com", "test1");

    expect(signInResult.data).toMatchSnapshot();
    expect(!!signInResult.data?.id).toBe(false);
    expect(signInResult.dataError?.error?.length).toBe(1);
    expect(signInResult.functionError).toBe(undefined);
    expect(signInResult.pending).toBe(false);

    expect(!!authAPI.user?.id).toBe(false);
    expect(authAPI.authenticated).toBe(false);
    expect(authAPI.loaded).toBe(true);
    expect(!!authAPI.token).toBe(false);
  });

  it("Can sign in", async () => {
    const signInResult = await authAPI.signIn("sdk@example.com", "test");

    expect(signInResult.data).toMatchSnapshot();
    expect(!!signInResult.data?.id).toBe(true);
    expect(signInResult.dataError).toBe(undefined);
    expect(signInResult.functionError).toBe(undefined);
    expect(signInResult.pending).toBe(false);

    expect(!!authAPI.user?.id).toBe(true);
    expect(authAPI.authenticated).toBe(true);
    expect(authAPI.loaded).toBe(true);
    expect(!!authAPI.token).toBe(true);
  });

  it("Does not cache mutations", async done => {
    await authAPI
      .signIn("sdk@example.com", "test")
      .then(() =>
        apiProxy.setPassword({
          email: "sdk@example.com",
          password: "test2345678",
          token: "5hr-73a06b70fd6ad8ab3913",
        })
      )
      .catch(() =>
        apiProxy.setPasswordChange({
          newPassword: "test2345678",
          oldPassword: "test2345678",
        })
      )
      .catch(() => {
        const extractedCache = cache.extract();
        expect(extractedCache.ROOT_MUTATION).toBeUndefined();

        done();
      });
  });

  it("Can sign out", async () => {
    expect(!!authAPI.user?.id).toBe(true);
    expect(authAPI.authenticated).toBe(true);
    expect(authAPI.loaded).toBe(true);
    expect(!!authAPI.token).toBe(true);

    const signOutResult = await authAPI.signOut();

    expect(signOutResult.pending).toBe(false);

    expect(!!authAPI.user?.id).toBe(false);
    expect(authAPI.authenticated).toBe(false);
    expect(authAPI.loaded).toBe(true);
    expect(!!authAPI.token).toBe(false);
  });
});
