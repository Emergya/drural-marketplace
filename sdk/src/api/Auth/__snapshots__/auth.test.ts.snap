// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth API Can sign in 1`] = `
Object {
  "__typename": "User",
  "addresses": Array [],
  "defaultBillingAddress": null,
  "defaultShippingAddress": null,
  "email": "sdk@example.com",
  "firstName": "",
  "id": "VXNlcjo2OA==",
  "isStaff": true,
  "lastName": "",
}
`;

exports[`Auth API Returns error if credentials are invalid 1`] = `undefined`;
